# motif-mark

motif-mark takes a fasta file and returns a visualization of each gene/intron, exon/cassette, and motif present.

To run motif-mark use this command:

```
path/motif-mark-oop.py -f path/to/fasta/input/file.fa -m /path/to/input/motif/text/file.txt
```

Example command:
```
./motif-mark-oop.py -f User/username/motif-mark/Figure_1.fasta -m User/username/motif-mark/Fig_1_motifs.txt
```

Example output of motif-mark

![Figure_1 Fasta Motif Mark Visualization](https://github.com/pgenge/motif-mark/blob/main/Figure_1.png)
